### PR TITLE
[9.0][10.0][FIX] l10n_do: update CoA (for IFRS), taxes, fiscal positions and journals

### PR DESCRIPTION
**Pull request by gustavovalverde**
Description of the issue/feature this PR addresses:
- CoA non-compliant with Dominican Republic regulations
- CoA not standardized for most companies "real-life" use
- Uncomplete Taxes
- Wrongly configured Fiscal Positions
- Wrongly configured sequences and journals for regulator's needs

Current behavior before PR:
- Everything can be done, but it is not a production ready localization

Desired behavior after PR is merged:
- Easier implementation for most of the people
- Minor customization to be fully compliant with local regulations
- Standardized configuration to help other Odoo consultants and implementers

To comply with local and tax regulations:
- The whole Charts of Accounts was changed
- Most of the Journals has been preconfigured to comply with tax regulator
- All the regulatory taxes were added and configured with group taxes

To standardize the Dominican implementations:
- An IFRS compatible CoA that does not need major changes for most companies
- A well structured CoA with a codification logic.
- Use of the Odoo original flow for creating tax compliant invoice sequences
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Yenthe666** [comment]
@mart-e and @qdp-odoo can you guys look at this?

**Comment by gustavovalverde** [comment]
Hi @mart-e and @qdp-odoo. This will help a lot for Odoo massification in our Country, please let me know if there's something that needs to be fixed before merging. Thnx 😄

**Comment by Yenthe666** [comment]
@mart-e second ping, the guy is curious about feedback! Sorry for another ping :)

**Comment by gustavovalverde** [comment]
Hi! Any updates?

**Comment by gustavovalverde** [comment]
Hi @mart-e, before keeping with the Localization update, I'll like to know if I'm on the right path with this changes. Sorry for the pings, I just need some feedback.

**Comment by qdp-odoo** [comment]
Hi @gustavovalverde ,

I just had a quick look: you're not supposed to create 
* account.journal, ir.sequence objects: You have journals that are created automatically for you when you install a CoA. With the journals, internal sequences are created as well (based on the short code) (and that's the same with other objects like stock.picking for example, so you really don't need to do that)
* account.account.type: you must reuse the ones defined in account module that allows you use the generic reports we made.
* ir.model.data: that's very ugly. Remove that please :-)

keep up the good work!

**Comment by gustavovalverde** [comment]
Thanks for the feedback @qdp-odoo! 👍 

I'm just curious about somehing:

- When you install a new CoA, Odoo configures a few journals: Bank, Sales, Misc, Exhange, etc. But in our country the Sales Journal, Purchase and a few others **must** have an specific numeration for regulatory purposes. What I want to do here is to overwrite those journals and use the ones with the sequences needed by our regulator. _Is this something that you would accept in a PR_? Or with each installation people have to make those configurations each time? 

- I changed the some additional sequences so they could be in Spanish. Is this a bad practice?

I'll be waiting for your feedback before doing any additional changes.

**Comment by gustavovalverde** [comment]
Ok @qdp-odoo, I've deleted the account.account.type. 

I'll hold the changes on the journal to keep up with the regulator discussions, so I can complete the changes and have a ready-to-pull PR. 

Thanks for the feedback so far 😄

**Comment by qdp-odoo** [comment]
@gustavovalverde I'd suggest that you deal with the sequences by using the translations instead: Each journal's default code can be translated so there's no point in having the journals and sequences record in your xml files.

**Comment by qdp-odoo** [comment]
btw, why are you still commiting in that PR? I though it was ready to be reviewed...

**Comment by gustavovalverde** [comment]
@qdp-odoo but, what I we're needing for the regulator is some journals for sales and purchases, each with an specific sequence numbering that the regulator uses for invoices. I'll do as you just suggested about the codes in transifex  (for the default journals). But we need to add 6 more journals with the regulator sequence. 

Can this be accomplished through a PR like this or should this be done using a third-party module? Just consider that I'm doing this to have a 100% compliant regulation, if it not possible I'll just segregate

**Comment by gustavovalverde** [comment]
*ups, miss clicked the wrong button*

@qdp-odoo  ... If this is not possible I'll just segregate the journal functions in a module in the market. 

Please let me know if this is the best approach.

**Comment by gustavovalverde** [comment]
@qdp-odoo  Now I understand why you complained about me adding journals; this is because of a bigger issue: https://github.com/odoo/odoo/issues/14059. 

I'll be doing the changes you asked for, considering that the Journal is not the best approach to give documents different sequences (even though is the only approach available in the core modules).

**Comment by gustavovalverde** [comment]
We're good to go 👍 @mart-e and @qdp-odoo

**Comment by gustavovalverde** [comment]
Guys, @mart-e and @qdp-odoo, is there anything else that has to be done here?

Sorry for the pings.

**Comment by Yenthe666** [comment]
@qdp-odoo 😉

**Comment by gustavovalverde** [comment]
Hi @qdp-odoo,
For:
> for the journals/sequencies: you can inherit the method _prepare_all_journals() in l10n_do in order to create the journals you want. I won't accept any ir.sequence object created through xml files ;-)

I'll leave that part untouched for now. I'll make it in another PR, so I can test the changes I'll like to do. When doing this, can I also overwrite the default journals or just add additional ones?

For:
> About the account.account.tag: modifying the XML-IDs would break the compatibility with the enterprise version (l10n_do_reports)... which is not acceptable. You must keep them untouched.

I don't get this one. Should I use the old _account_account_tag.xml_ file completly (and change the **field name**) or can I just reuse those old **record id**s?

**Comment by qdp-odoo** [comment]
> When doing this, can I also overwrite the default journals or just add additional ones?

if you need to overwrite the defaults: your code should test the country of the company, if it's DO then do your stuff (including the possible overwrite of the defaults), otherwise: call super(). 

if you need to complete the defaults: your code should call super(), then test the country of the company, if it's DO then do append stuff. 

> Should I use the old account_account_tag.xml file completly (and change the field name) or can I just reuse those old record ids?

You can rename the tags, but dont modify the xml-ids, and don't add or remove any.

**Comment by gustavovalverde** [comment]
> You can rename the tags, but dont modify the xml-ids, and **don't add or remove any**.

Well, that's a bummer as we have a few  missing there, and others that are not needed at all. I'll just keep the old file, but I'm actually on the side that both (community and enterprise) should be changed to meet the most common cases.

**Comment by gustavovalverde** [comment]
Ok @qdp-odoo, so I have change most of the things you've recommended; I even used the method _prepare_all_journals() A last commit after your feedback should be good to go. 

Actual Issues:
1. Even though I changed and reviewed the Journal codes translations (INV, BILL, MISC, EXCH) in transifex, none of them are being translated. Maybe @mart-e can have a look at this.

2. Those Enterprise report tags must be changed, they are practically useless if not implemented they way I have them in this CoA.

3. I would like to assign one of my accounts to the extra journal (not let it create one), but I can't get this to work, a little help on how to do it would be great as this new journal is taking the first liquidity account sequence, which I don't want to happend.

**Comment by mart-e** [comment]
> Even though I changed and reviewed the Journal codes translations (INV, BILL, MISC, EXCH) in transifex, none of them are being translated. Maybe @mart-e can have a look at this.

Which terms are you talking about? I am not sure how this PR is related to translations issues (you can contact me directly on transifex, account mart_e, if you want to discuss a transifex issue).

**Comment by qdp-odoo** [comment]
@gustavovalverde Please have a look at these 2 commits:
https://github.com/odoo/odoo/commit/a20d1e9256d8471201c4a41897ba4baa22b43fef
https://github.com/odoo/odoo/commit/3fc879e3

The first one will allow to define your own 'undistributed profits/losses' account, to avoid a hack to set the good code. 
The second one will help to define the extra cash journal with a higher sequence. Just overwrite the _get_default_bank_account_ids() function to add yours. and it should do the trick

**Comment by gustavovalverde** [comment]
> Which terms are you talking about? I am not sure how this PR is related to translations issues (you can contact me directly on transifex, account mart_e, if you want to discuss a transifex issue).

@mart-e, I already created the issue in Transifex (Odoo 9.0, account module). I made this changes based on @qdp-odoo recommendation: https://github.com/odoo/odoo/pull/13787#issuecomment-256940037

> Please have a look at these 2 commits:
> a20d1e9
> 3fc879e

@qdp-odoo Just to be sure, do you want me to overwrite my functions with this code, or just delete my function and wait for this commit to be applied on 9.0? (I know it will work on 10.0, but I also want this to work on 9.0)

**Comment by gustavovalverde** [comment]
Well, this commit https://github.com/odoo/odoo/commit/a20d1e9256d8471201c4a41897ba4baa22b43fef didn't work out for me. I actually had to keep the approach overwriting the one being created. 

Using the approach in your commit, it always created the 99999 account. Even though I tested it using the whole function.

**Comment by qdp-odoo** [comment]
ho I didn't notice it was for 9.0... I backported them in 7cdd90e and 87db2cc

For the account 99999: it will not be created _if_ you define an account.template of type "current year earnings' in the Chart of Accounts.

**Comment by gustavovalverde** [comment]
Ok @qdp-odoo, I think we're good to go.

**Comment by gustavovalverde** [comment]
Hi @qdp-odoo , just pinging to know if there's anything else to be done for this to be merged.

**Comment by edsersolis** [comment]
Guys if you could please give some priority to this PR it would be great and we would be very thankful. It has been a long run for us on this matter and we know that you are very busy but DR needs these improvements in the CoA asap. All changes have been done as suggested.

Thanks for all the help and guidance you've provided so far. @qdp-odoo @Yenthe666 sorry to bother so much on this.

**Comment by gustavovalverde** [comment]
Hi @qdp-odoo and @Yenthe666. 

I don't want to be annoying, but is there anything else that has to be done for this to be merged? Or maybe you guys are on vacation and @mart-e can give us a hand here 👍

**Comment by qdp-odoo** [comment]
you're not forgotten but we're pretty busy here, trying to finish everything before the next saas release. I'll have a look on this, but probably not before january 2017.

sorry for the delay

**Comment by gustavovalverde** [comment]
Too bad 😞

Well, we'll just start the community implementations in the clients that need to begging January with Odoo, we will assume that this will be merged just as is.

Regards.

**Comment by Yenthe666** [comment]
I've assigned you on it, that way it'll be easier to keep track of it.

**Comment by gustavovalverde** [comment]
Hi guys, and happy new year.

Just pinging here to know if this PR will be merged this week @Yenthe666 @qdp-odoo 

Regards.

File: addons/l10n_do/models/chart_template.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    @api.model
+    def _get_default_bank_account_ids(self):
+        return [
+            {'acc_name': _('Cash'), 'account_type': 'cash'},
+            {'acc_name': _('Caja Chica'), 'account_type': 'cash'},
+            {'acc_name': _('Bank'), 'account_type': 'bank'}
+            ]
+
+    @api.model
+    def default_get(self, fields):
```
Comment (qdp-odoo) [inline addons/l10n_do/models/chart_template.py]: this is not needed in order to create the right bank journals: only the above method is enough, so this method can be removed

File: addons/l10n_do/models/chart_template.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+from openerp import models, api, _
+
+
+class WizardMultiChartsAccounts(models.TransientModel):
+    _inherit = 'wizard.multi.charts.accounts'
+
+    @api.model
+    def _get_default_bank_account_ids(self):
+        return [
```
Comment (qdp-odoo) [inline addons/l10n_do/models/chart_template.py]: I'd rather do:
```python
if self.env.user.company_id.country.code.upper() == 'EC':
    return [ your stuff ]
return super()._get_default_bank_account_ids()
```
That way, we ensure the behavior won't change if your in a multi-company/multi-localization environment

File: addons/l10n_do/data/account_account_tag.xml [hunk 3]
```diff
@@ -1,99 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<odoo>
-    <data noupdate="1">
-
-
-    <!-- Account Tags -->
-
-        <record id="tag_R_ISR_10" model="account.account.tag">
```
Comment (qdp-odoo) [inline addons/l10n_do/data/account_account_tag.xml]: As I said: 
> You can rename the tags, but dont modify the xml-ids, and don't add or remove any.
The problem is: if you do that, you may break people current installation and force them to make an update of the module.

> Well, that's a bummer as we have a few missing there, and others that are not needed at all. I'll like to keep the old approach, but I'm actually on the side that both (community and enterprise) should be changed to meet the most common and practical cases.
You can do so, but then the 2 PR need to target master. And both the PR have to be done/merged together

So there are 2 solutions only: 
1) don't touch the xml-ids of the tags, just rename the field name if you want. This PR can be accepted in stable
2) do whatever you wanna do with the tags: you need to modify both the community and enterprise repo. We will only accept that in master if both PR are ready together

Comment (gustavovalverde) [inline addons/l10n_do/data/account_account_tag.xml]: How can I access the Enterprise repo to make this changes?

**Comment by gustavovalverde** [comment]
It should be ready now @qdp-odoo, please let me know.

**Comment by gustavovalverde** [comment]
After being pulled, will this be immediately ported to v10 and the Enterprise repo?

**Comment by RafosRD** [comment]
Is the best Dominican location, the community needs it

**Comment by eneldoserrata** [comment]
Nice job @gustavovalverde this update of the module works like a charm.

**Comment by qdp-odoo** [comment]
1) don't remove the menuitem l10n_do.account_reports_do_statements_menu as it is used in l10n_do_reports

2) can you try to reduce the diff by 
* keeping the tags definition in addons/l10n_do/data/account_account_tag.xml, in the same order
* renaming account_chart_template.xml into coa_template.xml instead of delete+create
* renaming set_account_on_chart_template.xml into account_defaults.xml instead of delete+create

thanks, it's already looking almost acceptable :-)

**Comment by qdp-odoo** [comment]
3) my former comment about _get_default_bank_account_ids() is still not done too

**Comment by juhenriquez** [comment]
Good Job @gustavovalverde this is looking real nice.

**Comment by jlopezg** [comment]
Good job! @gustavovalverde

**Comment by gustavovalverde** [comment]
> don't remove the menuitem l10n_do.account_reports_do_statements_menu as it is used in l10n_do_reports

Ok, I'll keep it, I removed it because I saw some new l10n_* changes that did the same for v10.


> can you try to reduce the diff by

keeping the tags definition in addons/l10n_do/data/account_account_tag.xml, in the same order
renaming account_chart_template.xml into coa_template.xml instead of delete+create
renaming set_account_on_chart_template.xml into account_defaults.xml instead of delete+create

**Comment by gustavovalverde** [comment]
> 2. can you try to reduce the diff by
> keeping the tags definition in addons/l10n_do/data/account_account_tag.xml, in the same order
> renaming account_chart_template.xml into coa_template.xml instead of delete+create
> renaming set_account_on_chart_template.xml into account_defaults.xml instead of delete+create

Is this necessary to be pulled? This is a hard Refactoring for me. 

> 3. my former comment about _get_default_bank_account_ids() is still not done too

I did the changes just as recommended here: https://github.com/odoo/odoo/pull/13787/files/b8a4a883b4aee67737638916fc037b82b4e30c8a#r96368207

That's done here: https://github.com/gustavovalverde/odoo/blob/a19566fe7c31992edf69dd7a1200cd380d3ad5e2/addons/l10n_do/models/chart_template.py

File: addons/l10n_do/models/chart_template.py (hunk 2 above)
Comment (gustavovalverde) [inline addons/l10n_do/models/chart_template.py]: Done: https://github.com/gustavovalverde/odoo/blob/a19566fe7c31992edf69dd7a1200cd380d3ad5e2/addons/l10n_do/models/chart_template.py

**Comment by gustavovalverde** [comment]
Hi @qdp-odoo, I made the renames based on the NEW API and considering this commit https://github.com/odoo/odoo/commit/ad7ca14d1b8efb5c1a4d574012bb2492914fb446

**Comment by gustavovalverde** [comment]
@qdp-odoo I'll be unavailable for a few days, Is there anything else to be done to this PR before it can be pulled? Our community is needing this changes.

File: addons/l10n_do/models/chart_template.py (hunk 2 above)
Comment (gustavovalverde) [inline addons/l10n_do/models/chart_template.py]: It actually was `if self.env.user.company_id.country.code_id.upper()` "_id" was missing.

**Comment by gustavovalverde** [comment]
@qdp-odoo Is there a way to also pre-configure the _tax_group_id_ in each tax? I have the groups, but I can't find a way to preassign them in the tax template

**Comment by caudata** [comment]
This is MUST-HAVE addition to Odoo regionalization for the Dominican Republic. I'm quite sure that this could be the top feature that Odoo needs to finally enter into massive "Odoofication" in this nation; specially since it's now struggling with half-baked,slow-to-evolve, proprietary solutions; all unaffordable for average SMBs and entrepeneurs alike. Cheers!

**Comment by gustavovalverde** [comment]
Guys @Yenthe666 @qdp-odoo, any updates?

**Comment by gustavovalverde** [comment]
Ping @Yenthe666 @qdp-odoo. 

Sorry if I'm being a bit stubborn guys, but we really need your acceptance before going into production and we have a bunch of projects stopped waiting for this. cc: @mart-e

**Comment by Yenthe666** [comment]
@qdp-odoo there is quite a bit on hold. Perhaps you can give the guy feedback again?

**Comment by gustavovalverde** [comment]
Updates?

**Comment by qdp-odoo** [comment]
man, I understand this is quite critical for you but poking us every damn day will not speed up things. We are all pretty busy and we prioritize our tasks as much as we can... it just appears than we are not fully devoted to the Dominican Republic localization, and personally, finding time to check your PR is not on my priority list. 

Don't misunderstand me: I already found time to check it and I will continue to do so. Your contribution is really appreciated and I don't minimize its importance, but everything important is not urgent... and your insistence is driving me nuts.

**Comment by gustavovalverde** [comment]
I'm sorry, but I have no way to know the amount of people working as reviewers behind this Github repo, nor your priorities, nor your workload.

A feedback can even be: "I'll be looking at this the next month", and I'll wait a month if it's needed...but If I don't have that knowledge or the feedback, I'll ping.

I suggest you guys have public priorities dashboard, so we can be aware of what's urgent for you and for the community...and what's not. Or even have public SLAs and assign priority based on the subject or how it affects Odoo (overall).

Communication is the key.

**Comment by edsersolis** [comment]
@qdp-odoo pinging you hasn't been on a daily basis. If you had specified the fixes you wanted to have on this all at once it would have been done in a week, but we already have 4 months trying to get this pulled and you keep asking for small fixes everytime, and now you complain of being pinged "daily"?! 

As said at first, we know you may be busy, it may not your priority but it is ours and have put a lot of effort to get this done. If you don't have the time or don't care as you just stated, why don't you please assign another person for this?

**Comment by Yenthe666** [comment]
Hi guys,

I believe that both parties have valid points.
@gustavovalverde I agree that the transparency is too low. I personally don't work for Odoo but have some extra rights here and I can reach some people easier but I still face this problem too.
@edsersolis I agree with you too. I can clearly see that you've guys put quite some work in this and I'd like to thank you for this. Without people like you it would takes ages to get this.
However, that being said, it is also hard for Odoo. Don't forget that @qdp-odoo is an employee and has responsibility towards other people and tasks. Simply assigning somebody else isn't easy either as the resources are limited and he is probably by far the best accounting guy at Odoo.

So, we all have valid points but we can't solve this issue overnight. @qdp-odoo perhaps you can just put a date on it so other people have a time table and so that you don't get pinged anymore?
I will have a talk with Fabien, Martin and Olivier to see if we can do anything to improve these kind of cases.

Regards,
Yenthe

**Comment by gustavovalverde** [comment]
Hi guys @Yenthe666 @qdp-odoo.

Any updates on this? We're needing this pull so we can keep up with the changes on the enterprise git.

Regards.

**Comment by Yenthe666** [comment]
Hi @gustavovalverde,

I'm sorry but I can't do much more, this one is fully at qdp-odoo but I'd rather not ping him too much (we already made him mad I guess 😢)
Let us give him some time I guess.

Regards,
Yenthe

**Comment by qdp-odoo** [comment]
This PR has been merged in revision 06c5db50b6212497514fa7e3706efafd872ed10b

Hooray :-)

Thanks for your contribution and your patience!

**Comment by gustavovalverde** [comment]
Love you! ❤️ 😛

**Comment by edsersolis** [comment]
Yay! Finally this will be available with ease.

Thanks!

**Comment by Yenthe666** [comment]
Thanks a lot @qdp-odoo !  Great job!

**Pull request closed without merging**